File: census-app/frontend/src/lib/analytics.ts
/** Local analytics.
 *
 * Computed from whatever the device holds so supervisors and administrators get
 * charts even with no connectivity. When a backend is available the same shape
 * comes back from `/api/analytics/summary` over the full dataset.
 */

import type { AnalyticsSummary, Household, User, Zone } from './types';
import { percent, todayKey } from './utils';

export const AGE_BANDS: Array<{ key: string; min: number; max: number }> = [
  { key: 'band.0_6', min: 0, max: 6 },
  { key: 'band.7_14', min: 7, max: 14 },
  { key: 'band.15_29', min: 15, max: 29 },
  { key: 'band.30_44', min: 30, max: 44 },
  { key: 'band.45_59', min: 45, max: 59 },
  { key: 'band.60_plus', min: 60, max: Number.POSITIVE_INFINITY },
];

function bandFor(age: number): string {
  const band = AGE_BANDS.find((candidate) => age >= candidate.min && age <= candidate.max);
  return band ? band.key : 'band.60_plus';
}

export function computeAnalytics(
  households: Household[],
  zones: Zone[] = [],
  users: User[] = [],
): AnalyticsSummary {
  const genderCounts: Record<string, number> = { male: 0, female: 0, transgender: 0 };
  const ageBands: Record<string, number> = {};
  const casteCounts: Record<string, number> = {};
  const religionCounts: Record<string, number> = {};
  const literacy = { literate: 0, illiterate: 0 };
  const dailyMap = new Map<string, number>();

  for (const band of AGE_BANDS) ageBands[band.key] = 0;

  let members = 0;
  let submitted = 0;
  let approved = 0;
  let flagged = 0;
  let drafts = 0;

  for (const household of households) {
    if (household.status === 'submitted') submitted += 1;
    else if (household.status === 'approved') approved += 1;
    else if (household.status === 'flagged') flagged += 1;
    else drafts += 1;

    const day = todayKey(household.submittedAt ?? household.createdAt);
    if (day) dailyMap.set(day, (dailyMap.get(day) ?? 0) + 1);

    for (const member of household.members) {
      members += 1;
      if (member.gender) genderCounts[member.gender] = (genderCounts[member.gender] ?? 0) + 1;
      if (member.age !== null && member.age !== undefined && member.age >= 0) {
        const key = bandFor(member.age);
        ageBands[key] = (ageBands[key] ?? 0) + 1;
      }
      if (member.casteCategory) {
        casteCounts[member.casteCategory] = (casteCounts[member.casteCategory] ?? 0) + 1;
      }
      if (member.religion) {
        religionCounts[member.religion] = (religionCounts[member.religion] ?? 0) + 1;
      }
      if (member.literate === true) literacy.literate += 1;
      else if (member.literate === false) literacy.illiterate += 1;
    }
  }

  /* Zones ----------------------------------------------------------- */
  const zoneProgress = zones.map((zone) => {
    const collected = households.filter((household) => household.zoneId === zone.id).length;
    return {
      zoneId: zone.id,
      zoneName: zone.name,
      collected,
      target: zone.targetHouseholds,
      coverage: percent(collected, zone.targetHouseholds),
    };
  });

  // Households in zones the device does not know about still deserve a row.
  const knownZoneIds = new Set(zones.map((zone) => zone.id));
  const orphanCount = households.filter(
    (household) => !household.zoneId || !knownZoneIds.has(household.zoneId),
  ).length;
  if (orphanCount) {
    zoneProgress.push({
      zoneId: '',
      zoneName: 'Unassigned',
      collected: orphanCount,
      target: 0,
      coverage: 0,
    });
  }

  /* Enumerators ------------------------------------------------------ */
  const byEnumerator = new Map<string, AnalyticsSummary['enumeratorProgress'][number]>();
  for (const household of households) {
    const id = household.enumeratorId || 'unknown';
    let entry = byEnumerator.get(id);
    if (!entry) {
      const user = users.find((candidate) => candidate.id === id);
      entry = {
        enumeratorId: id,
        name: household.enumeratorName || user?.name || 'Unknown',
        collected: 0,
        submitted: 0,
        approved: 0,
        flagged: 0,
      };
      byEnumerator.set(id, entry);
    }
    entry.collected += 1;
    if (household.status === 'submitted') entry.submitted += 1;
    if (household.status === 'approved') entry.approved += 1;
    if (household.status === 'flagged') entry.flagged += 1;
    if (!entry.lastActivity || entry.lastActivity < household.updatedAt) {
      entry.lastActivity = household.updatedAt;
    }
  }

  /* Daily counts — last 14 days, including zero days so the chart is honest */
  const dailyCounts: Array<{ date: string; count: number }> = [];
  const today = new Date();
  for (let offset = 13; offset >= 0; offset -= 1) {
    const date = new Date(today);
    date.setDate(today.getDate() - offset);
    const key = todayKey(date);
    dailyCounts.push({ date: key, count: dailyMap.get(key) ?? 0 });
  }

  const targetHouseholds = zones.reduce((total, zone) => total + zone.targetHouseholds, 0);

  return {
    households: households.length,
    members,
    submitted,
    approved,
    flagged,
    drafts,
    targetHouseholds,
    coverage: percent(households.length, targetHouseholds),
    averageHouseholdSize: households.length
      ? Math.round((members / households.length) * 10) / 10
      : 0,
    genderCounts,
    ageBands,
    literacy,
    casteCounts,
    religionCounts,
    zoneProgress: zoneProgress.sort((a, b) => b.collected - a.collected),
    enumeratorProgress: Array.from(byEnumerator.values()).sort((a, b) => b.collected - a.collected),
    dailyCounts,
  };
}

/** Females per 1000 males, the standard census presentation. */
export function sexRatio(genderCounts: Record<string, number>): number | null {
  const males = genderCounts.male ?? 0;
  const females = genderCounts.female ?? 0;
  if (!males) return null;
  return Math.round((females / males) * 1000);
}

/** Households a supervisor should look at first. */
export function needsAttention(households: Household[]): Household[] {
  return households.filter((household) => {
    if (household.status === 'flagged') return true;
    if (household.status === 'draft') return false;
    if (household.flags.some((flag) => flag.severity === 'error')) return true;
    if (!household.location && household.collectedBy === 'enumerator') return true;
    return false;
  });
}

File: census-app/backend/app/analytics.py
"""Server-side analytics — the same shape the frontend computes locally."""

from __future__ import annotations

from collections import defaultdict
from datetime import date, datetime, timedelta, timezone
from typing import Any

AGE_BANDS: list[tuple[str, int, float]] = [
    ("band.0_6", 0, 6),
    ("band.7_14", 7, 14),
    ("band.15_29", 15, 29),
    ("band.30_44", 30, 44),
    ("band.45_59", 45, 59),
    ("band.60_plus", 60, float("inf")),
]


def _band_for(age: int) -> str:
    for key, low, high in AGE_BANDS:
        if low <= age <= high:
            return key
    return "band.60_plus"


def _day_key(value: str | None) -> str:
    if not value:
        return ""
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return value[:10] if len(value) >= 10 else ""
    return parsed.astimezone(timezone.utc).date().isoformat()


def _percent(part: int, total: int) -> float:
    if not total:
        return 0.0
    return round(part / total * 1000) / 10


def compute_summary(
    households: list[dict[str, Any]],
    zones: list[dict[str, Any]],
    users: list[dict[str, Any]],
) -> dict[str, Any]:
    gender_counts: dict[str, int] = {"male": 0, "female": 0, "transgender": 0}
    age_bands: dict[str, int] = {key: 0 for key, _, _ in AGE_BANDS}
    caste_counts: dict[str, int] = defaultdict(int)
    religion_counts: dict[str, int] = defaultdict(int)
    literacy = {"literate": 0, "illiterate": 0}
    daily: dict[str, int] = defaultdict(int)

    members = 0
    submitted = approved = flagged = drafts = 0

    for household in households:
        status = household.get("status", "draft")
        if status == "submitted":
            submitted += 1
        elif status == "approved":
            approved += 1
        elif status == "flagged":
            flagged += 1
        else:
            drafts += 1

        day = _day_key(household.get("submittedAt") or household.get("createdAt"))
        if day:
            daily[day] += 1

        for member in household.get("members", []):
            members += 1
            gender = member.get("gender")
            if gender:
                gender_counts[gender] = gender_counts.get(gender, 0) + 1

            age = member.get("age")
            if isinstance(age, int) and age >= 0:
                age_bands[_band_for(age)] += 1

            if member.get("casteCategory"):
                caste_counts[member["casteCategory"]] += 1
            if member.get("religion"):
                religion_counts[member["religion"]] += 1

            literate = member.get("literate")
            if literate is True:
                literacy["literate"] += 1
            elif literate is False:
                literacy["illiterate"] += 1

    # Zones -----------------------------------------------------------
    zone_progress = []
    known_zone_ids = set()
    for zone in zones:
        known_zone_ids.add(zone.get("id"))
        collected = sum(1 for h in households if h.get("zoneId") == zone.get("id"))
        target = int(zone.get("targetHouseholds", 0) or 0)
        zone_progress.append(
            {
                "zoneId": zone.get("id", ""),
                "zoneName": zone.get("name", ""),
                "collected": collected,
                "target": target,
                "coverage": _percent(collected, target),
            }
        )

    orphans = sum(1 for h in households if h.get("zoneId") not in known_zone_ids)
    if orphans:
        zone_progress.append(
            {
                "zoneId": "",
                "zoneName": "Unassigned",
                "collected": orphans,
                "target": 0,
                "coverage": 0.0,
            }
        )

    # Enumerators ------------------------------------------------------
    by_name = {user.get("id"): user.get("name", "") for user in users}
    progress: dict[str, dict[str, Any]] = {}
    for household in households:
        key = household.get("enumeratorId") or "unknown"
        entry = progress.setdefault(
            key,
            {
                "enumeratorId": key,
                "name": household.get("enumeratorName") or by_name.get(key) or "Unknown",
                "collected": 0,
                "submitted": 0,
                "approved": 0,
                "flagged": 0,
                "lastActivity": None,
            },
        )
        entry["collected"] += 1
        status = household.get("status")
        if status in ("submitted", "approved", "flagged"):
            entry[status] += 1
        updated = household.get("updatedAt")
        if updated and (entry["lastActivity"] is None or updated > entry["lastActivity"]):
            entry["lastActivity"] = updated

    # Last 14 days, zero-filled so the chart does not lie about quiet days.
    today = date.today()
    daily_counts = []
    for offset in range(13, -1, -1):
        key = (today - timedelta(days=offset)).isoformat()
        daily_counts.append({"date": key, "count": daily.get(key, 0)})

    target_total = sum(int(zone.get("targetHouseholds", 0) or 0) for zone in zones)

    return {
        "households": len(households),
        "members": members,
        "submitted": submitted,
        "approved": approved,
        "flagged": flagged,
        "drafts": drafts,
        "targetHouseholds": target_total,
        "coverage": _percent(len(households), target_total),
        "averageHouseholdSize": (
            round(members / len(households) * 10) / 10 if households else 0.0
        ),
        "genderCounts": dict(gender_counts),
        "ageBands": dict(age_bands),
        "literacy": literacy,
        "casteCounts": dict(caste_counts),
        "religionCounts": dict(religion_counts),
        "zoneProgress": sorted(zone_progress, key=lambda row: row["collected"], reverse=True),
        "enumeratorProgress": sorted(
            progress.values(), key=lambda row: row["collected"], reverse=True
        ),
        "dailyCounts": daily_counts,
    }
